### ISSUE TITLE
trenni: invalid operands to binary expression ('VALUE' (aka 'void *') and 'int')

### ISSUE THREAD

**Issue by deepj** [oracle/truffleruby#1605]
To reproduce:

```
gem install trenni
```

Error:

```
Fetching trenni-3.6.1.gem
Fetching rake-compiler-1.0.7.gem
Successfully installed rake-compiler-1.0.7
Building native extensions. This could take a while...
ERROR:  Error installing trenni:
	ERROR: Failed to build gem native extension.

    current directory: ~/.gem/truffleruby/2.4.4/gems/trenni-3.6.1/ext
~/.rubies/truffleruby-1.0.0-rc13/bin/truffleruby -rrubygems ~/.gem/truffleruby/2.4.4/gems/rake-12.3.2/exe/rake RUBYARCHDIR\=~/.gem/truffleruby/2.4.4/extensions/x86_64-darwin/2.4.0/trenni-3.6.1 RUBYLIBDIR\=~/.gem/truffleruby/2.4.4/extensions/x86_64-darwin/2.4.0/trenni-3.6.1
fatal: not a git repository (or any of the parent directories): .git
mkdir -p tmp/x86_64-darwin/trenni/2.4.4
cd tmp/x86_64-darwin/trenni/2.4.4
~/.rubies/truffleruby-1.0.0-rc13/bin/truffleruby -I. ../../../../trenni/extconf.rb
checking for rb_sym2str()... yes
checking for rb_str_cat_cstr()... yes
checking for rb_str_reserve()... no
creating Makefile
cd -
cd tmp/x86_64-darwin/trenni/2.4.4
/usr/bin/make
compiling ../../../../trenni/escape.c
../../../../trenni/escape.c:6:6: error: invalid operands to binary expression ('VALUE' (aka 'void *') and 'int')
        if (RB_IMMEDIATE_P(value))
            ^~~~~~~~~~~~~~~~~~~~~
~/.rubies/truffleruby-1.0.0-rc13/lib/cext/include/ruby/ruby.h:331:39: note: expanded from macro 'RB_IMMEDIATE_P'
#define RB_IMMEDIATE_P(x) ((VALUE)(x) & RUBY_IMMEDIATE_MASK)
                           ~~~~~~~~~~ ^ ~~~~~~~~~~~~~~~~~~~
1 error generated.
make: *** [escape.bc] Error 1
rake aborted!
Command failed with status (2): [/usr/bin/make...]
~/.gem/truffleruby/2.4.4/gems/rake-compiler-1.0.7/lib/rake/extensiontask.rb:166:in `block (2 levels) in define_compile_tasks'
~/.gem/truffleruby/2.4.4/gems/rake-compiler-1.0.7/lib/rake/extensiontask.rb:165:in `block in define_compile_tasks'
~/.gem/truffleruby/2.4.4/gems/rake-12.3.2/exe/rake:27:in `<main>'
Tasks: TOP => default => compile => compile:x86_64-darwin => compile:trenni:x86_64-darwin => copy:trenni:x86_64-darwin:2.4.4 => tmp/x86_64-darwin/trenni/2.4.4/trenni.su
(See full trace by running task with --trace)

rake failed, exit code 1

Gem files will remain installed in ~/.gem/truffleruby/2.4.4/gems/trenni-3.6.1 for inspection.
Results logged to ~/.gem/truffleruby/2.4.4/extensions/x86_64-darwin/2.4.0/trenni-3.6.1/gem_make.out
```

**Comment by aardvark179** [comment]
Thank you for reporting this. I think we will need to alter this macros slightly as our ’VALUE’s are pointers rather than longs.

**Comment by Xenofex** [comment]
Similar problem here. `VALUE` is now a `void *` in TruffleRuby, while in MRI, it is a `uintptr_t`. As a result of which the `switch` statement is broken here:

https://github.com/k0kubun/hamlit/blob/94b14839cd42e8b70715ba7c0caf18c2c47ec67a/ext/hamlit/hamlit.c#L287

```c
    switch (value) { // value being a VALUE
      case Qtrue:
        rb_str_concat(buf, str_space());
        rb_str_concat(buf, key);
        break;
      case Qnil:
        break; // noop
      case Qfalse:
        break; // noop
      default:
        rb_str_concat(buf, str_space());
        rb_str_concat(buf, key);
        rb_str_concat(buf, str_equal());
        rb_str_concat(buf, quote);
        rb_str_concat(buf, escape_attribute(escape_attrs, to_s(value)));
        rb_str_concat(buf, quote);
        break;
    }
  }
```

**Comment by chrisseaton** [comment]
Yes, this is an incompatibility we have, documented here.

https://github.com/oracle/truffleruby/blob/f5bde7ebe118380413d1711acb1f205880eeea64/doc/user/compatibility.md#value-is-a-pointer

The reason for this is that we don't want to convert managed objects to have handles, which we have to do in order to produce a real numerical value (our `void*` values don't have a numerical value like a native pointer.) The biggest problem with this is that we can't switch on these values.

**Comment by deepj** [comment]
@chrisseaton So, there is some workaround or any solution how the maintainers make their gems requiring this to be compatible with TruffleRuby?

**Comment by chrisseaton** [comment]
Gem maintainers can switch to using `if` statements if they want to, but we're not asking anyone to do that yet and we're still thinking about a better solution. Theoretical options include a custom LLVM front-end to rewrite them automatically, for example.

**Comment by Xenofex** [comment]
That would be awesome. Or just some friendlier error message with a link to what to do and why.

**Comment by deepj** [comment]
@eregon this can be closed since this issue is solved in master

**Comment by eregon** [comment]
Indeed, I fixed this in 0552ac72e3396492dc42ea53c1ac5ee2eb24a3ba but missed to close this issue, thanks for checking.
